File: projects/typescript/06-event-emitter/starter/main.ts
// 06-event-emitter (starter)
declare const require: any;
declare const process: any;
const readline = require('readline');

// TODO(step-1): EventMap 제네릭 타입
type Listener<T> = (data: T) => void;

// TODO(step-2): TypedEventEmitter 클래스
class TypedEventEmitter<Events extends Record<string, any>> {
  private listeners: { [K in keyof Events]?: Listener<Events[K]>[] } = {};

  // TODO(step-3): on 메서드
  on<K extends keyof Events>(event: K, listener: Listener<Events[K]>): void {}

  // TODO(step-4): emit 메서드
  emit<K extends keyof Events>(event: K, data: Events[K]): void {}
}

// TODO(step-5): REPL 실행기
function main(): void {
  const rl = readline.createInterface({ input: process.stdin });
  rl.on('line', () => {});
}

main();

File: projects/python/06-event-emitter/starter/main.py
# 06-event-emitter (starter)
import sys

# TODO(step-1): EventEmitter 클래스 구조
class EventEmitter:
    def __init__(self):
        self.listeners = {}

    # TODO(step-2): on (지속 구독)
    def on(self, event: str, callback):
        pass

    # TODO(step-3): once (일회성 구독)
    def once(self, event: str, callback):
        pass

    # TODO(step-4): emit (이벤트 발행)
    def emit(self, event: str, data: str):
        pass

# TODO(step-5): 명령줄 인터페이스
def main():
    pass

if __name__ == "__main__":
    main()
